core/shared/src/main/scala/zio/ZIOMetric.scala

```diff
@@ -582,14 +706,40 @@ object ZIOMetric {
      * Increments the counter associated with the specified value by one.
      */
     def observe(value: String)(implicit trace: ZTraceElement): UIO[Any] =
-      setCount.observe(value)
+      withSetCount(_.observe(value))
 
     /**
-     * Returns the number of occurences of every value observed by this set
+     * Returns the number of occurrences of every value observed by this set
      * count.
      */
     def occurrences(implicit trace: ZTraceElement): UIO[Chunk[(String, Long)]] =
-      setCount.occurrences
+      withSetCount(_.occurrences)
+
+    /**
+     * Converts this set count metric to one where the tags depend on the
+     * measured effect's result value
+     */
+    def taggedWith(f: A => Chunk[MetricLabel]): ZIOMetric[A] = {
+      if (self.setCountRef eq null) {
```

Comment:
I think this may violate referential transparency, because it has a side-effect which is not described in the return type. 

Will the behaviour of the original `self` differ in any way after this method is invoked?

Reply:
Thanks, this made me realize that there is such a problem: if `c1` is a counter and `c2` is the modified counter, applying `c2` first modifies the `FiberRef` and does not restore it, so a next call to `c1` on the same fiber will have wrong tags.

`taggedWith` should call `self.copy`

Reply:
Actually just checked and `copy` does not work correctly currently either (without my change):

```scala
  val count = ZIOMetric
    .count("test", MetricLabel("x", "1"))
    .copy("test2", Chunk(MetricLabel("x", "2")))

  val run =
    for {
      _      <- ZIO.succeed("hello") @@ count
      states <- UIO(MetricClient.unsafeStates)
      _      <- ZIO.debug(states.toList.mkString("\n"))
    } yield ()
```

prints

```
(Counter(test2,Chunk(MetricLabel(x,2))),MetricState(test2{x->2}, Counter(0.0)))
(Counter(test,Chunk(MetricLabel(x,1))),MetricState(test{x->1}, Counter(1.0)))
```

(because the copied aspect just calls the original `self`'s `apply` which refers to the original metric).

So in addition to fix the referential transparency issue, should also:
- fix `copy`
- write some metric tests :)